### PR TITLE
feat: Private API endpoint for open calls

### PR DESCRIPTION
**Pull request by martintomas**
I am adding new private API endpoint for OpenCalls. Endpoint itself is quite simple, but what made this task more challenging is searching ability, because:
 - country/municipality/department names are localized
 - name of open call is also localized (we should be able to deal with fallbacks)
 - instrument types is array of enums
 - status is enum
 
It was required to use full text search on all these attributes, BUT instrument_types or status are not texts and names are text, but there is extra logic which is enforced on app layer (fallback). I was able to come with three possible solutions:
 - do searching at app --> just load everything to app, run it through serializer and do searching then (simple, but quite slow for big sets and not really usable for backoffice)
 - modify PgSearch gem --> spend some time reading doc, but didn't find some nice way how to do it
 - use ransack --> backoffice already uses this and it is quite easy to extend, cons is that it is gonna be probably slower then PgSearch (seq scan/filter which it does looks quite intimidating :))
 
Whatever, I have decided to go for ransack option and written extra concern which adds three possible helpers to model:
- `generate_ransackers_for_translated_columns` --> for every translated columns add ransacker option which fallbacks same way as traco at app
- `generate_localized_ransackers_for_static_types` --> adds ransackers for static types. Allows to search or sort based on user language. For arrays, it does `array_to_string`
- `generate_localized_ransackers_for_enums` --> similar as previous, allows to search/sort based on status text which is localized.

## Testing instructions

Via Rswag documentation -- new API endpoint should be part of it, feel free to check all fulltext filters

## Tracking

https://vizzuality.atlassian.net/browse/LET-940

### CODE REVIEW

**Comment by tsubik** [comment]
Could we also do it like full_text OR contains in the rest of problematic fields?

File: backend/app/models/concerns/extra_ransackers.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+module ExtraRansackers
```
Comment (tsubik) [inline backend/app/models/concerns/extra_ransackers.rb:1]: :+1:

**Comment by martintomas** [comment]
> Could we also do it like full_text OR contains in the rest of problematic fields?

No sure about question @tsubik :). Basically you can use any matchers mentioned at https://activerecord-hackery.github.io/ransack/getting-started/search-matches/ and some of them should be quite close to full text search (I have used `cont`, but we can use any of them). True full text search is handled by `PgSearch` which we already have at project, but I had problem to enhance it so it works with enum, static types or translation fallbacks. You can ofc use combination --> full text for all string values at db which are not translated (don't have fallback) and ransacker for rest.

**Comment by tsubik** [comment]
I meant something like in pseudo code `OpenCall.full_text_search.or(OpenCall.ransack_contains_for_troubling_attributes_scope)`.

**Comment by tsubik** [comment]
But actually, I see there is no description, only name, so maybe `cont` is good enough as all of the searched fields are short strings.

**Comment by martintomas** [comment]
> I meant something like in pseudo code `OpenCall.full_text_search.or(OpenCall.ransack_contains_for_troubling_attributes_scope)`.

Aha, I get it, hmm, I know that it is possible to combine queries generated by `PgSearch ` and `ransacker`, but I had problems when joins come into play, because they use different aliases. My experience here was that it is possible, but also quite fragile and does not work for all use cases. If it would be required to search also on top of `description` (which is localized with traco), we could either use `PgSearch` and give up on fallbacks or use ransack which is for sure gonna be slower.

**Review by tsubik — APPROVED** [review]
:+1:

**Comment by tsubik** [comment]
Yeah, when combining complicated queries it's better to just `pluck(:id)` from one query and combine with simplest `where(id: ...)`, but here it's fine how it is.

**Comment by martintomas** [comment]
> Yeah, when combining complicated queries it's better to just `pluck(:id)` from one query and combine with simplest `where(id: ...)`, but here it's fine how it is.

Yes, exactly, with `pluck`, you can combine everything :D

**Review by agnessa — APPROVED** [review]

**Pull request merged**